### PR TITLE
Templated LDA.  Changed testing structure so that FLD and SVD cases weren't both being used in a single FLD or SVD test case.

### PR DESCRIPTION
**Pull request by c4goldsw**

### CODE REVIEW

**Comment by c4goldsw** [comment]
@vigsterkr Hey, would you be able to look over this when you get the chance to?

File: src/shogun/classifier/LDA.cpp [hunk 1]
```diff
@@ -19,6 +19,8 @@
 #include <shogun/mathematics/Math.h>
 #include <shogun/mathematics/eigen3.h>
 
+#include <iostream>
```
Comment (vigsterkr) [inline src/shogun/classifier/LDA.cpp]: do we really need iostream?

File: src/shogun/classifier/LDA.cpp [hunk 2]
```diff
@@ -95,45 +113,49 @@ bool CLDA::train_machine(CFeatures *data)
...
+	typename SGVector<ST>::EigenVectorXt mean_neg(num_feat);
+	mean_neg.setZero();
+	typename SGVector<ST>::EigenVectorXt mean_pos(num_feat);
+	mean_pos.setZero();
 
 	//mean neg
-	for(i=0; i<num_neg; i++)
-		mean_neg+=fmatrix.col(classidx_neg[i]);
-	mean_neg/=(float64_t)num_neg;
+	for(i=0; i<num_neg; i++){
```
Comment (vigsterkr) [inline src/shogun/classifier/LDA.cpp]: i would hope that there's an eigen matrix operator for doing this summation, or linalg by chance?

File: src/shogun/classifier/LDA.cpp [hunk 3]
```diff
@@ -216,22 +238,32 @@ bool CLDA::train_machine(CFeatures *data)
...
-		bias=(x.transpose()*mean_total);
+		bias= ((float64_t) (x.transpose()*mean_total));
 		bias=bias*(-1);
 	}
+
+ 	w=SGVector<float64_t>(num_feat);
+	w.zero();
+
+	//copy w_st into w
+	for(i = 0; i < w.size(); ++i)
```
Comment (vigsterkr) [inline src/shogun/classifier/LDA.cpp]: this definitely could be done simpler.. .like a memcpy? :)

File: src/shogun/classifier/LDA.cpp (hunk 1 above)
Comment (c4goldsw) [inline src/shogun/classifier/LDA.cpp]: My bad, I forgot to remove it.

File: src/shogun/classifier/LDA.cpp (hunk 3 above)
Comment (c4goldsw) [inline src/shogun/classifier/LDA.cpp]: How am I meant to do a memcpy if the elements are of different types?

File: src/shogun/classifier/LDA.cpp (hunk 2 above)
Comment (c4goldsw) [inline src/shogun/classifier/LDA.cpp]: I'll look into that.

Comment (c4goldsw) [inline src/shogun/classifier/LDA.cpp]: CDenseFeatures has subsets, (I'd have to look into that a little more), which we could use to set the values for ```mean_neg``` and ```mean_pos``` respectively.  But, when setting ```fmatrix```, since it's an eigen matrix, I'd still have to use a loop to set it's values according to the indexes in ```classidx_neg```.  If CDenseFeatures simply creates copies using loops, I think we should stick with loops.

File: src/shogun/classifier/LDA.cpp [hunk 4]
```diff
@@ -66,18 +68,34 @@ bool CLDA::train_machine(CFeatures *data)
 		if(!data->has_property(FP_DOT))
 			SG_ERROR("Specified features are not of type CDotFeatures\n")
 		set_features((CDotFeatures*) data);
+	} else {
```
Comment (karlnapf) [inline src/shogun/classifier/LDA.cpp]: please respect shoguns code style

File: src/shogun/classifier/LDA.cpp [hunk 5]
```diff
@@ -66,18 +68,34 @@ bool CLDA::train_machine(CFeatures *data)
...
+		data = get_features();
+		REQUIRE(data, "Features have not been provided!  Please pass in training features into train_machine() or into the constructor.\n")
```
Comment (karlnapf) [inline src/shogun/classifier/LDA.cpp]: Please dont mention method names or technical terms such as "constructor" in error messages for users.
Just say "Features have not been provided.\n"

File: src/shogun/classifier/LDA.cpp [hunk 6]
```diff
@@ -66,18 +68,34 @@ bool CLDA::train_machine(CFeatures *data)
...
 	}
 
-	REQUIRE(features, "Features are not provided!\n")
 	SGVector<int32_t>train_labels=((CBinaryLabels *)m_labels)->get_int_labels();
 	REQUIRE(train_labels.vector,"Provided Labels are empty!\n")
 
-	SGMatrix<float64_t>feature_matrix=((CDenseFeatures<float64_t>*)features)
+	REQUIRE(data->get_num_vectors() == train_labels.vlen,"Number of training examples(%d) should be "
+		"equal to number of labels specified(%d)!\n", data->get_num_vectors(), train_labels.vlen);
```
Comment (karlnapf) [inline src/shogun/classifier/LDA.cpp]: remove specified and add space before (%d)

File: src/shogun/classifier/LDA.cpp [hunk 7]
```diff
@@ -66,18 +68,34 @@ bool CLDA::train_machine(CFeatures *data)
...
+
+	if(data->get_feature_type() == F_SHORTREAL)
+		return CLDA::train_machine_templated<float32_t>(train_labels, data);
```
Comment (karlnapf) [inline src/shogun/classifier/LDA.cpp]: I wonder whether we can macro this somehow @vigsterkr ?

File: src/shogun/classifier/LDA.cpp [hunk 8]
```diff
@@ -95,45 +113,49 @@ bool CLDA::train_machine(CFeatures *data)
...
 	}
 
-	w=SGVector<float64_t>(num_feat);
-	w.zero();
-	MatrixXd fmatrix=Map<MatrixXd>(feature_matrix.matrix, num_feat, num_vec);
-	VectorXd mean_neg(num_feat);
-	mean_neg=VectorXd::Zero(num_feat);
-	VectorXd mean_pos(num_feat);
-	mean_pos=VectorXd::Zero(num_feat);
+	SGVector<ST> w_st(num_feat);
```
Comment (karlnapf) [inline src/shogun/classifier/LDA.cpp]: why the rename? This might be better in another patch

File: src/shogun/classifier/LDA.cpp (hunk 2 above)
Comment (karlnapf) [inline src/shogun/classifier/LDA.cpp]: newline before {

Comment (karlnapf) [inline src/shogun/classifier/LDA.cpp]: I think this is fine as this single loop is fast.  It would be more code slickness ... so maybe have a method?

File: src/shogun/classifier/LDA.cpp [hunk 9]
```diff
@@ -95,45 +113,49 @@ bool CLDA::train_machine(CFeatures *data)
...
 
 	//mean neg
-	for(i=0; i<num_neg; i++)
-		mean_neg+=fmatrix.col(classidx_neg[i]);
-	mean_neg/=(float64_t)num_neg;
+	for(i=0; i<num_neg; i++){
+		mean_neg+= fmatrix.col(classidx_neg[i]);
+	}
+
+	mean_neg/= (ST) num_neg;
```
Comment (karlnapf) [inline src/shogun/classifier/LDA.cpp]: space before /=

File: src/shogun/classifier/LDA.cpp [hunk 10]
```diff
@@ -95,45 +113,49 @@ bool CLDA::train_machine(CFeatures *data)
...
 
 	// get m(-ve) - mean(-ve)
 	for(i=0; i<num_neg; i++)
-		fmatrix.col(classidx_neg[i])-=mean_neg;
+		fmatrix.col(classidx_neg[i])-= mean_neg;
```
Comment (karlnapf) [inline src/shogun/classifier/LDA.cpp]: please dont do this. If you edit a file, stick to its style, if you wanna change the style, send a separate patch

File: src/shogun/classifier/LDA.cpp [hunk 11]
```diff
@@ -95,45 +113,49 @@ bool CLDA::train_machine(CFeatures *data)
...
 
 	// get m(+ve) - mean(+ve)
 	for(i=0; i<num_pos; i++)
-		fmatrix.col(classidx_pos[i])-=mean_pos;
+		fmatrix.col(classidx_pos[i])-= mean_pos;
 
-	SGMatrix<float64_t>scatter_matrix(num_feat, num_feat);
-	Map<MatrixXd> scatter(scatter_matrix.matrix, num_feat, num_feat);
+	SGMatrix<ST>scatter_matrix(num_feat, num_feat);
+	typename SGMatrix<ST>::EigenMatrixXtMap scatter(scatter_matrix.matrix, num_feat, num_feat);
```
Comment (karlnapf) [inline src/shogun/classifier/LDA.cpp]: it would be cool if we could have some snytatic sugar to make this shorter.
@lisitsyn ?

File: src/shogun/classifier/LDA.cpp [hunk 12]
```diff
@@ -95,45 +113,49 @@ bool CLDA::train_machine(CFeatures *data)
...
 		// covariance matrix.
-		MatrixXd cov_mat(num_feat, num_feat);
+		typename SGMatrix<ST>::EigenMatrixXt cov_mat(num_feat, num_feat);
 		cov_mat=fmatrix*fmatrix.transpose();
 		scatter=cov_mat/(num_vec-1);
-		float64_t trace=scatter.trace();
-		double s=1.0-m_gamma;
-		scatter *=s;
-		scatter.diagonal()+=VectorXd::Constant(num_feat, trace*m_gamma/num_feat);
+		ST trace=scatter.trace();
```
Comment (karlnapf) [inline src/shogun/classifier/LDA.cpp]: please dont use such methods on SGMatrix - they are obsolete.
Use linalg CMath or eigen

Comment (karlnapf) [inline src/shogun/classifier/LDA.cpp]: Best to add it to linalg and use that

File: src/shogun/classifier/LDA.cpp (hunk 3 above)
Comment (karlnapf) [inline src/shogun/classifier/LDA.cpp]: either that or leave it as is ....

File: src/shogun/classifier/LDA.h [hunk 13]
```diff
@@ -163,6 +163,9 @@ class CLDA : public CLinearMachine
 		 */
 		virtual bool train_machine(CFeatures* data=NULL);
 
+		template <typename ST>
+		bool train_machine_templated(SGVector<int32_t> train_labels, CFeatures * features);
```
Comment (karlnapf) [inline src/shogun/classifier/LDA.h]: doxygen

File: tests/unit/classifier/LDA_unittest.cc [hunk 14]
```diff
@@ -34,97 +34,111 @@
...
 	CBinaryLabels* labels;
 	CDenseFeatures<float64_t>* features;
+
+	generate_test_data<float64_t>(labels, features);
+
+	//Train the binary classifier.
+	CLDA lda_FLD(0, features, labels, FLD_LDA);
+	lda_FLD.train();
+
+	//Test.
```
Comment (karlnapf) [inline tests/unit/classifier/LDA_unittest.cc]: superflous comment

File: tests/unit/classifier/LDA_unittest.cc [hunk 15]
```diff
@@ -34,97 +34,111 @@
...
+
+template <typename ST>
+void FLD_test(SGVector<ST> &projection_FLD, SGVector<ST> &w_FLD)
+{
 	CBinaryLabels* labels;
 	CDenseFeatures<float64_t>* features;
+
+	generate_test_data<float64_t>(labels, features);
+
+	//Train the binary classifier.
```
Comment (karlnapf) [inline tests/unit/classifier/LDA_unittest.cc]: superflous comment -- the next method is called train ... no need for that :)

File: tests/unit/classifier/LDA_unittest.cc [hunk 16]
```diff
@@ -176,3 +206,35 @@ TEST_F(LDATest, DISABLED_CheckProjection_SVD)
 	EXPECT_NEAR(+6.90668279, projection_SVD[8], epsilon);
 	EXPECT_NEAR(+7.96084156, projection_SVD[9], epsilon);
 }
+
+//FLD template testing
+TEST(LDA, FLD_template_test_float32)
```
Comment (karlnapf) [inline tests/unit/classifier/LDA_unittest.cc]: Can we use a macro here?
Or doenst work with the googletest macros?

File: src/shogun/classifier/LDA.cpp (hunk 8 above)
Comment (c4goldsw) [inline src/shogun/classifier/LDA.cpp]: CLinearMachine contains a weights variable```SGVector<float64_t> w```.  I need a templated variable to store the weights during the training before I copy the final weight vector back into ```w``` ([line 265](https://github.com/shogun-toolbox/shogun/pull/3480/files#diff-7c03379ee9137c8e8719907bd06da7c2R265)).  If I were just to use ```SGVector<ST> w``` inside of the code, I'd get shadowing warnings.  So, I think it's better just to use ```w_st```.

File: src/shogun/classifier/LDA.cpp (hunk 9 above)
Comment (c4goldsw) [inline src/shogun/classifier/LDA.cpp]: Style question: should I also do this for anything of the form [operator]= (e.g. mean_neg+=)?

File: tests/unit/classifier/LDA_unittest.cc (hunk 16 above)
Comment (c4goldsw) [inline tests/unit/classifier/LDA_unittest.cc]: Hmm, I don't understand what the macro should do - should it compact the below into a single line?

```
TEST(LDA, FLD_template_test_floatmax)
{
	check_eigenvectors_fld<floatmax_t>();
}
```

File: src/shogun/classifier/LDA.cpp (hunk 2 above)
Comment (c4goldsw) [inline src/shogun/classifier/LDA.cpp]: I'll implement this in a separate PR.

File: src/shogun/classifier/LDA.cpp (hunk 3 above)
Comment (c4goldsw) [inline src/shogun/classifier/LDA.cpp]: I'll implement this in a separate PR, since it requires a class/struct which contains the transformation operation.  It would be a good utility method.

File: src/shogun/classifier/LDA.cpp (hunk 12 above)
Comment (c4goldsw) [inline src/shogun/classifier/LDA.cpp]: I'm a bit confused - ```scatter``` is an an eigen map, so ```trace()``` is eigen's trace method (```scatter_matrix``` is the SGMatrix that is wrapped by ```scatter```).

File: src/shogun/classifier/LDA.cpp (hunk 10 above)
Comment (c4goldsw) [inline src/shogun/classifier/LDA.cpp]: Another style question: if there are inconsistencies with whitespace usage from a previous PR, should I remove the whitespaces?

File: src/shogun/classifier/LDA.cpp (hunk 3 above)
Comment (karlnapf) [inline src/shogun/classifier/LDA.cpp]: yeah definitely good utitlity method. Linalg could do this

File: src/shogun/classifier/LDA.cpp (hunk 2 above)
Comment (karlnapf) [inline src/shogun/classifier/LDA.cpp]: yep!

**Comment by c4goldsw** [comment]
@vigsterkr Cleaned up.

File: src/shogun/classifier/LDA.cpp [hunk 17]
```diff
@@ -175,38 +192,38 @@ bool CLDA::train_machine(CFeatures *data)
...
 		//within class matrix
-		MatrixXd Sw= num_pos*cen_pos+num_neg*cen_neg;
-		float64_t trace=Sw.trace();
-		double s=1.0-m_gamma;
+		typename SGMatrix<ST>::EigenMatrixXt Sw= num_pos*cen_pos+num_neg*cen_neg;
+		ST trace=Sw.trace();
+		ST s=1.0-((ST)m_gamma);
 		Sw *=s;
-		Sw.diagonal()+=VectorXd::Constant(num_feat, trace*m_gamma/num_feat);
+		Sw.diagonal()+=Eigen::DenseBase<typename SGVector<ST>::EigenVectorXt>::Constant(num_feat, trace*((ST)m_gamma)/num_feat);
```
Comment (c4goldsw) [inline src/shogun/classifier/LDA.cpp]: @vigsterkr As far as I know, there is no VectorXt in Eigen - it'd be nice to have that type.  I can create that in another PR.

File: tests/unit/classifier/LDA_unittest.cc (hunk 16 above)
Comment (karlnapf) [inline tests/unit/classifier/LDA_unittest.cc]: Nevermind actually, this is fine!

File: src/shogun/classifier/LDA.cpp (hunk 8 above)
Comment (karlnapf) [inline src/shogun/classifier/LDA.cpp]: I get it! Good! The name could be better, or a source code comment would help

**Comment by karlnapf** [comment]
I think this is fine. Let's merge and check the buildbot, etc.

Can you also send a patch that updates the NEWS?
